src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingGraphInterface.java

```diff
@@ -0,0 +1,1102 @@
+        }
+
+        // the first element must be an M
+        if ( requireFirstElementM && elements.get(0).getOperator() != CigarOperator.M ) {
+            return false;
+        }
+
+        // the last element must be an M
+        if ( requireLastElementM && elements.get(numElements - 1).getOperator() != CigarOperator.M ) {
+            return false;
+        }
+
+        // note that there are checks for too many mismatches in the dangling branch later in the process
+
+        return true;
+    }
+
+    /**
+     * calculates the longest suffix match between a sequence and a smaller kmer
+     *
+     * @param seq         the (reference) sequence
+     * @param kmer        the smaller kmer sequence
+     * @param seqStart    the index (inclusive) on seq to start looking backwards from
+     * @return the longest matching suffix
+     */
+    @VisibleForTesting
+    static int longestSuffixMatch(final byte[] seq, final byte[] kmer, final int seqStart) {
+        for ( int len = 1; len <= kmer.length; len++ ) {
+            final int seqI = seqStart - len + 1;
+            final int kmerI = kmer.length - len;
+            if ( seqI < 0 || seq[seqI] != kmer[kmerI] ) {
+                return len - 1;
+            }
+        }
+        return kmer.length;
+    }
+
+    @VisibleForTesting
+    protected void setAlreadyBuilt() {
+        alreadyBuilt = true;
+    }
+
+    @VisibleForTesting
+    void setMaxMismatchesInDanglingHead(final int maxMismatchesInDanglingHead) {
+        this.maxMismatchesInDanglingHead = maxMismatchesInDanglingHead;
+    }
+
+    /**
+     * Return the collection of outgoing vertices that expand this vertex with a particular base.
+     *
+     * @param v original vertex.
+     * @param b expanding base.
+     * @return never null, but perhaps an empty set. You cannot assume that you can modify the result.
+     */
+    @VisibleForTesting
+    Set<MultiDeBruijnVertex> getNextVertices(final MultiDeBruijnVertex v, final byte b) {
+        Utils.nonNull(v, "the input vertex cannot be null");
+        Utils.validateArg(vertexSet().contains(v), "the vertex must be present in the graph");
+        final List<MultiDeBruijnVertex> result = new LinkedList<>();
+        for (final MultiDeBruijnVertex w : outgoingVerticesOf(v)) {
+            if (w.getSuffix() == b) {
+                result.add(w);
+            }
+        }
+        switch (result.size()) {
+            case 0: return Collections.emptySet();
+            case 1: return Collections.singleton(result.get(0));
+            default:
+                    return new HashSet<>(result);
+        }
+    }
+
+    /**
+     * Find vertex and its position in seqForKmers where we should start assembling seqForKmers
+     *
+     * @param seqForKmers the sequence we want to thread into the graph
+     * @return the position of the starting vertex in seqForKmer, or -1 if it cannot find one
+     */
+    protected int findStart(final SequenceForKmers seqForKmers) {
```

Comment:
I'm actually in favor of the find start returning -1 rather than an optional. If you think about what a SequenceForKmers is, it is effectively a list and we are just finding the index in that list corresponding to the first valid position on the graph. All of the java graph code is predicated on -1 being the "failed to find" index.

Reply:
That's fair.  I might lobby for changing all of our graph code, but that's my hill to die on, not yours.